### ISSUE TITLE
Integrate SWOT satellite geometry into dMC routing to break equifinality

### ISSUE THREAD

**Issue by taddyb** [DeepGroundwater/ddr#159]
## Motivation

The dMC model suffers from equifinality when learning channel geometry: the gauge loss only constrains velocity, so multiple (n, p, top_width) combinations produce identical routing. The SWOT satellite (Cerbelaud et al., Nature 2026) provides observed width-elevation relationships for 126,674 reaches globally, including ~11,000 in North America. By fixing top_width and side_slope from SWOT observations for covered reaches (and falling back to KAN prediction for uncovered headwaters), we break the equifinality manifold from a surface to a line (only n remains free).

The existing code already supports this pattern — Lynker hydrofabric loads TopWdth/ChSlp into `routing_dataclass`, and the `numel() == 0` check in `_denormalize_spatial_parameters` selects between data and KAN. We extend this to support **partial coverage** (SWOT for some reaches, KAN for others) via NaN-masked blending.

## Data Pipeline

```
SWOT EIV_fits CSVs ──► references/build_swot_geometry.py ──► data/swot_merit_geometry.nc
(SWORD reach_id)        + MERIT-SWORD crosswalk                (COMID dimension)
                                                                     │
                                                        merit.py loads at init
                                                                     │
                                                        routing_dataclass.top_width
                                                        routing_dataclass.side_slope
                                                        (NaN where no SWOT)
                                                                     │
                                                        mmc.py blends: SWOT where
                                                        valid, KAN where NaN
```

## SWOT → Model Parameter Derivation

From EIV_fits per SWORD reach:
- **top_width** = `med_width` (median observed width, meters)
- **side_slope** = `m_1 / 2` (in-channel dW/dH slope ÷ 2 → H:V ratio), clamped to [0.5, 50.0]

---

### Step 1: Preprocessing Script (NEW file)

**File**: `references/geo_io/build_swot_geometry.py`

Standalone CLI script following the pattern of `references/geo_io/build_gage_references.py`:

1. Read all EIV_fits CSVs from PFAF regions 71-78 (North America)
   - Source: `/mnt/ssd1/data/swot/swot-river-volume/output_revision/output_revision/EIV_fits/`
   - Key columns: `reach_id`, `med_width`, `m_1`, `nobs`, `fit_method`, `med_flow_area`
2. Read MERIT-SWORD crosswalk NetCDFs (`mb_to_sword` files in `ms_translate.zip`)
   - Source: `/mnt/ssd1/data/swot/` (Zenodo 14675925)
   - Maps MERIT COMID → SWORD reach_id
3. Join: for each MERIT COMID with a SWORD match that has EIV data:
   - `top_width = clamp(med_width, 1.0, 5000.0)`
   - `side_slope = clamp(m_1 / 2, 0.5, 50.0)`
4. Save as NetCDF with `COMID` dimension, variables: `top_width`, `side_slope`, `nobs`, `med_flow_area`
   - NaN for reaches without SWOT coverage

### Step 2: Config Schema Change

**File**: `src/ddr/validation/configs.py` — `DataSources` class

Add one field:
```python
swot_geometry: Path | None = Field(
    default=None,
    description="Path to preprocessed SWOT geometry NetCDF (COMID-indexed top_width + side_slope)",
)
```

Backward compatible: defaults to None, existing configs unchanged.

### Step 3: MERIT Geodataset Changes

**File**: `src/ddr/geodatazoo/merit.py`

**3a**: Add `_load_swot_geometry()` method in `__init__` that opens the .nc and builds a dict `{comid: (top_width, side_slope)}` for non-NaN entries. Only loads if `cfg.data_sources.swot_geometry` is set.

**3b**: Modify `_build_common_tensors` — if SWOT geometry loaded, build tensors sized `(n_segments,)` with SWOT values where available and NaN elsewhere. Fall through to `torch.empty(0)` if no SWOT data configured.

### Step 4: Routing Engine Changes

**File**: `src/ddr/routing/mmc.py` — `_denormalize_spatial_parameters`

Extract a `_resolve_parameter` helper that handles three cases:
1. `numel() == 0` → pure KAN (current MERIT path, unchanged)
2. No NaN in tensor → pure data (Lynker path, unchanged)
3. Has NaN → `torch.where(~nan_mask, data, kan_denormalized)` (NEW: SWOT partial coverage)

**Gradient flow**: `torch.where` passes gradients through the KAN branch for NaN-masked elements. SWOT values are fixed constants (no grad). KAN learns for uncovered reaches, SWOT constrains covered ones.

---

## Files Modified

| File | Change | Risk |
|------|--------|------|
| `references/geo_io/build_swot_geometry.py` | NEW ~150 lines | None — standalone preprocessing |
| `src/ddr/validation/configs.py` | +4 lines (optional field) | Low — backward compatible |
| `src/ddr/geodatazoo/merit.py` | +30 lines (load + populate) | Medium — data pipeline |
| `src/ddr/routing/mmc.py` | Refactor ~50 lines | Medium — routing core |

**Lynker path**: Completely unaffected (never produces empty tensors, no SWOT config).
**MERIT without SWOT**: Unaffected (`swot_geometry=None` → empty tensors → pure KAN).

## Concerns

1. **MERIT-SWORD crosswalk quality**: The crosswalk uses geometric overlap, not hydrological matching. Short or braided reaches may get mismatched SWORD geometry.
   - Mitigation: `nobs` field in the .nc allows post-hoc filtering (require nobs ≥ 10)

2. **Geometry discontinuity at SWOT/KAN boundary**: Adjacent reaches may have SWOT=500m and KAN=50m.
   - Mitigation: Less harmful than current equifinality. KAN learns from routed discharge backprop.

3. **m_1 ≈ 0 for rectangular fits**: Some SWOT reaches have near-zero dW/dH (rectangular channels).
   - Mitigation: Clamped to side_slope minimum of 0.5. These reaches are physically rectangular-like.

4. **SWOT data still downloading**: SWORD v16 and MERIT-SWORD crosswalk are incomplete. The preprocessing script can't run until downloads finish. The model code changes are independent and can proceed now.

## Assumptions

- `m_1 / 2` is a valid side_slope derivation (trapezoidal approximation of the in-channel W(H) slope)
- Primary SWORD match (`sword_1` field in crosswalk) is sufficient (no multi-match averaging)
- SWOT geometry from Oct 2023–Sep 2024 is representative of long-term channel shape (geomorphology is stable)
- CONUS PFAF regions 71-78 cover the model's domain

## Verification

1. Run preprocessing: `python references/geo_io/build_swot_geometry.py --eiv-fits-dir ... --crosswalk-zip ... --conus-adjacency ... --output data/swot_merit_geometry.nc`
2. Inspect output: check coverage %, top_width/side_slope distributions, NaN count
3. Run training with `swot_geometry: data/swot_merit_geometry.nc` in config
4. Compare learned depths to expected values (rivers should not be unrealistically shallow)
5. Compare routing metrics (KGE/NSE at gauges) with and without SWOT geometry
